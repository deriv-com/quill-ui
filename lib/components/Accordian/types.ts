import { ReactNode } from "react";
import { AccordionVariants } from ".";
import { TRegularSizes } from "@types";

export type AccordionProps = {
    id?: string;
    className?: string;
    title?: string;
    subtitle?: string;
    icon?: ReactNode;
    color?: string;
    textSize?: TRegularSizes;
    expanded?: boolean;
    divider?: AccordionDivider;
    disabled?: boolean;
    contentClassname?: string;
    expandedColor?: TExpandedColor;
    content?: () => ReactNode;
    customContent?: () => ReactNode;
    isFlush?: boolean;
    onExpand?: (e: boolean, i: string) => void;
};
type TExpandedColor = "white" | "gray";
export interface AccordionTabProps extends Omit<AccordionProps, "content"> {
    tab?: Tab;
    className?: string;
    variant?: keyof AccordionVariants;
    content: {
        className?: string;
        data: AccordionProps[][];
    };
    multiCollapse?: boolean;
}

export type Tab = {
    align?: TabAlign;
    data: TabData[];
};

export type TabData = {
    id: number;
    title: string;
};

export type TabAlign = "center" | "end";

export type AccordionDivider = "none" | "both" | "bottom";
