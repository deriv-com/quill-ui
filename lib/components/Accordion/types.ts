import { ReactNode } from "react";
import { AccordionVariants } from ".";
import { TRegularSizes } from "@types";

export type AccordionProps = {
    id?: string;
    className?: string;
    title?: string;
    subtitle?: string;
    icon?: ReactNode;
    color?: string;
    textSize?: TRegularSizes;
    expanded?: boolean;
    divider?: AccordionDivider;
    disabled?: boolean;
    contentClassname?: string;
    content?: () => ReactNode;
    customContent?: () => ReactNode;
    onExpand?: (e: boolean, i: string) => void;
};

export interface AccordionTabProps extends Omit<AccordionProps, "content"> {
    className?: string;
    variant?: keyof AccordionVariants;
    content: {
        className?: string;
        data: AccordionProps[][];
    };
    multiCollapse?: boolean;
}

export type AccordionDivider = "none" | "both" | "bottom";
