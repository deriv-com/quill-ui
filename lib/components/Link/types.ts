import { TAllSizesExceptExtraSmall } from "@types";
import { LinkHTMLAttributes, ReactNode } from "react";

export type LinkSize = "caption" | TAllSizesExceptExtraSmall;
export type TLinkColor = "black" | "white";

export interface LinkProps extends LinkHTMLAttributes<HTMLAnchorElement> {
    className?: string;
    size?: LinkSize;
    color?: TLinkColor;
    disabled?: boolean;
    icon?: ReactNode;
    hasChevron?: boolean;
    target?: "_blank" | "_self" | "_parent" | "_top";
    children?: React.ReactNode;
}
