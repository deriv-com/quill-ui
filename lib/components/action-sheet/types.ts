import { ComponentPropsWithoutRef } from "react";

type OpenType =
    | {
          isOpen?: boolean;
          onOpen: () => void;
      }
    | {
          isOpen?: undefined;
          onOpen?: () => void;
      };

export type RootProps = ComponentPropsWithoutRef<"div"> &
    OpenType & {
        onClose?: () => void;
        type?: "modal" | "non-modal";
        expandable?: boolean;
    };

type ActionType = {
    content: string;
    onAction: () => void;
};

export type FooterProps = ComponentPropsWithoutRef<"div"> & {
    primaryAction?: ActionType;
    secondaryAction?: ActionType;
};

export type FooterAlignment = FooterProps;
