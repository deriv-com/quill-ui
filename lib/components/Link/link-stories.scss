.link-stories {
    width: 150px;
    padding: var(--core-spacing-800);
    border-radius: var(--core-borderRadius-200);
    background-color: var(--component-textIcon-normal-prominent);
}