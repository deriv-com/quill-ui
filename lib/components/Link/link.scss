.link {
    display: inline-flex;
    gap: 8px;
    text-decoration: none;
    cursor: pointer;
    align-items: center;
    &__color {
        &-black {
            color: var(--core-color-solid-slate-1400);
        }
        &-white {
            color: var(--core-color-solid-slate-50);
        }
    }
    color: var(--component-textIcon-normal-prominent);

    &:hover {
        text-decoration: underline;
    }

    &:active {
        text-decoration: underline;
    }

    &:disabled {
        pointer-events: none;
        opacity: var(--core-opacity-600);
    }

    &__icon-color {
        &-black {
            fill: var(--core-color-solid-slate-1400);
        }
        &-white {
            fill: var(--core-color-solid-slate-50);
        }
    }

    &__icon-size {
        &-caption {
            width: 8px;
            height: 18px;
        }

        &-sm {
            width: 8px;
            height: 22px;
        }

        &-md {
            width: 10px;
            height: 24px;
        }

        &-lg {
            width: 13px;
            height: 30px;
        }

        &-xl {
            width: 15px;
            height: 36px;
        }
    }

}