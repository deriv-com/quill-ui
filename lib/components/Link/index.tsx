import { forwardRef } from "react";
import { LabelPairedChevronRightSmRegularIcon } from "@deriv/quill-icons/LabelPaired";
import { LinkProps } from "./types";
import clsx from "clsx";
import "./link.scss";
import { CaptionText, Text } from "@components/Typography";

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(
    (
        {
            className,
            size = "md",
            disabled,
            color = "black",
            icon,
            hasHoverEffect = true,
            hasChevron = false,
            children,
            target = "_blank",
            href,
            ...rest
        },
        ref,
    ) => {
        return (
            <a
                ref={ref}
                target={target}
                href={disabled ? "javascript:void(0)" : href}
                {...rest}
                className={clsx(
                    "quill-link",
                    `quill-link__color-${color}`,
                    disabled && "quill-link--disabled",
                    hasHoverEffect && "quill-link--hover",
                    className,
                )}
            >
                {icon && (
                    <span data-testid="dt-link-icon" className="quill-link__icon">
                        {icon}
                    </span>
                )}
                {size === "caption" ? (
                    <CaptionText color={color}>{children}</CaptionText>
                ) : (
                    <Text
                        color={color}
                        size={
                            size === "sm"
                                ? "sm"
                                : size === "lg"
                                  ? "lg"
                                  : size === "xl"
                                    ? "xl"
                                    : "md"
                        }
                    >
                        {children}
                    </Text>
                )}

                {hasChevron && (
                    <LabelPairedChevronRightSmRegularIcon
                        data-testid="dt-link-chevron"
                        className={clsx(
                            `quill-link__icon-color-${color}`,
                            `quill-link__icon-size-${size}`,
                        )}
                    />
                )}
            </a>
        );
    },
);

Link.displayName = "Link";

export default Link;
