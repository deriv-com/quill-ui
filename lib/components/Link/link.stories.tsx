import type { Meta, StoryObj } from "@storybook/react";
import Link from ".";
import {
    LabelPairedPlaceholderCaptionRegularIcon,
    LabelPairedPlaceholderLgRegularIcon,
    LabelPairedPlaceholderMdRegularIcon,
    LabelPairedPlaceholderSmRegularIcon,
    LabelPairedPlaceholderXlRegularIcon,
} from "@deriv/quill-icons/LabelPaired";
import { ComponentProps } from "react";

const meta = {
    title: "Components/Links/Link",
    component: Link,
    parameters: {
        layout: "centered",
    },
    args: {
        size: "md",
    },
    argTypes: {
        size: {
            options: ["caption", "sm", "md", "lg", "xl"],
            control: { type: "radio" },
            description: "To select the size of the link",
        },
        color: {
            options: ["black", "white"],
            control: { type: "radio" },
            description: "To select the color of the link",
        },
    },
    tags: ["autodocs"],
} satisfies Meta<typeof Link>;

export default meta;
type Story = StoryObj<typeof meta>;
const placeholder = {
    caption: <LabelPairedPlaceholderCaptionRegularIcon />,
    sm: <LabelPairedPlaceholderSmRegularIcon />,
    md: <LabelPairedPlaceholderMdRegularIcon />,
    lg: <LabelPairedPlaceholderLgRegularIcon />,
    xl: <LabelPairedPlaceholderXlRegularIcon />,
};
export const DefaultBlackLink: Story = {
    args: {
        children: "link",
        size: "md",
        disabled: false,
        hasChevron: true,
        color: "black",
    },
};
export const DefaultWhiteLink: Story = {
    args: {
        children: "link",
        size: "md",
        color: "white",
        disabled: false,
        hasChevron: true,
    },
};

export const LinkWithIconBlack = (args: ComponentProps<typeof Link>) => (
    <Link
        {...args}
        children="link"
        color="black"
        icon={placeholder[args?.size as keyof typeof placeholder]}
    />
);
export const LinkWithIconWhite = (args: ComponentProps<typeof Link>) => (
    <Link
        {...args}
        children="link"
        size="md"
        color="white"
        icon={<LabelPairedPlaceholderMdRegularIcon fill="#ffffff" />}
    />
);
