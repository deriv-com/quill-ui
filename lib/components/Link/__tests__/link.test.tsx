import { render, screen } from "@testing-library/react";
import Link from "../index";
import { LabelPairedPlaceholderMdRegularIcon } from "@deriv/quill-icons/LabelPaired";
import { LinkSize, TLinkColor } from "../types";

describe("Link", () => {
    it("should render correctly", () => {
        render(<Link>Click me</Link>);
        const LinkElement = screen.getByText(/click me/i);
        expect(LinkElement).toBeInTheDocument();
    });

    it("should render chevron icon when hasChevron prop is true", () => {
        render(<Link hasChevron>Click me</Link>);
        const icon = screen.getByTestId("dt-link-chevron");
        expect(icon).toBeInTheDocument();
    });

    it("should render icon when a icon is passed", () => {
        render(
            <Link icon={<LabelPairedPlaceholderMdRegularIcon />}>
                Click me
            </Link>,
        );
        const icon = screen.getByTestId("dt-link-icon");
        expect(icon).toBeInTheDocument();
    });

    const colors: TLinkColor[] = ["black", "white"];
    const sizes: LinkSize[] = ["caption", "sm", "md", "lg", "xl"];

    colors.forEach((color) => {
        sizes.forEach((size) => {
            it(`should render correctly with ${size} size and color ${color}`, () => {
                const { container } = render(
                    <Link size={size} color={color}>
                        Click me - {size}
                    </Link>,
                );
                expect(container).toMatchSnapshot();
            });
        });
    });

    const disabled = [true, false];
    disabled.forEach((dis) => {
        it(`should render correctly with disabled ${dis}`, () => {
            render(<Link disabled={dis}>Click me disabled</Link>);
            const LinkElement = screen.getByText("Click me disabled");
            expect(LinkElement).toMatchSnapshot();
        });
    });
});
