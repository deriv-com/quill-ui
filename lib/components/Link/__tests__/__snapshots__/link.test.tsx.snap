// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render correctly with caption size and color black 1`] = `
<div>
  <a
    class="quill-link quill-link__color-black quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__caption-text__weight--regular__decoration--default black"
    >
      Click me - 
      caption
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with caption size and color white 1`] = `
<div>
  <a
    class="quill-link quill-link__color-white quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__caption-text__weight--regular__decoration--default white"
    >
      Click me - 
      caption
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with disabled false 1`] = `
<p
  class="quill-typography quill-typography__body-text__size--md__weight--regular__decoration--default black"
>
  Click me disabled
</p>
`;

exports[`Link should render correctly with disabled true 1`] = `
<p
  class="quill-typography quill-typography__body-text__size--md__weight--regular__decoration--default black"
>
  Click me disabled
</p>
`;

exports[`Link should render correctly with lg size and color black 1`] = `
<div>
  <a
    class="quill-link quill-link__color-black quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--lg__weight--regular__decoration--default black"
    >
      Click me - 
      lg
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with lg size and color white 1`] = `
<div>
  <a
    class="quill-link quill-link__color-white quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--lg__weight--regular__decoration--default white"
    >
      Click me - 
      lg
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with md size and color black 1`] = `
<div>
  <a
    class="quill-link quill-link__color-black quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--md__weight--regular__decoration--default black"
    >
      Click me - 
      md
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with md size and color white 1`] = `
<div>
  <a
    class="quill-link quill-link__color-white quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--md__weight--regular__decoration--default white"
    >
      Click me - 
      md
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with sm size and color black 1`] = `
<div>
  <a
    class="quill-link quill-link__color-black quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--sm__weight--regular__decoration--default black"
    >
      Click me - 
      sm
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with sm size and color white 1`] = `
<div>
  <a
    class="quill-link quill-link__color-white quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--sm__weight--regular__decoration--default white"
    >
      Click me - 
      sm
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with xl size and color black 1`] = `
<div>
  <a
    class="quill-link quill-link__color-black quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--xl__weight--regular__decoration--default black"
    >
      Click me - 
      xl
    </p>
  </a>
</div>
`;

exports[`Link should render correctly with xl size and color white 1`] = `
<div>
  <a
    class="quill-link quill-link__color-white quill-link--hover"
    target="_blank"
  >
    <p
      class="quill-typography quill-typography__body-text__size--xl__weight--regular__decoration--default white"
    >
      Click me - 
      xl
    </p>
  </a>
</div>
`;
